Check alternate index during stale data check

MartinRiese opened the pull request:
## Product Description

Only affects the `stale_data_in_es` command

## Technical Summary

https://dimagi.atlassian.net/browse/USH-4625

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

MartinRiese committed:
Check alternate index during stale data check

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -154,19 +163,36 @@ class CaseHelper:
     def run(cls, run_config):
-        for chunk in cls.get_sql_chunks(run_config):
-            yield from cls._yield_missing_in_es(chunk)
+        domains_with_dedicated_index = list(settings.CASE_SEARCH_SUB_INDICES.keys())
```
I think you mentioned looking at doing this one domain at a time, is that right?  That does sound like it'd end up a lot simpler, though I haven't thought it all through.  We're already pulling in chunks of 1000 at a time, so adding a breakpoint in that for each domain doesn't sound like it'd turn this into something that struggles to run at scale.

MartinRiese replied:
The chunks returned by `get_sql_chunks` are not sorted by domain. So each chunk of 1000 could have any number of different domains. Adding that order by seemed unnecessary if there is at most one domain with a dedicated domain and in the case or swiss, india or self hosted instances none.

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -175,43 +201,65 @@ def get_sql_chunks(run_config):
 
     @staticmethod
-    def _yield_missing_in_es(chunk):
+    def _yield_missing_in_es(chunk, dedicated_case_search_index=None):
```
The existing version of this operates on a chunk that might contain cases that should be present in the `case_search` index and some that shouldn't be (that's what `_get_ids_from_chunk` does).  I don't think that logic is relevant anymore, if all domains are on the case search index.  Still, now we have two separate ways of splitting the logic for two separate per-domain settings, which is kind of weird.  Have you looked at simplifications or doing the same logic for both?

MartinRiese replied:
are you referring to `if domain_needs_search_index(domain):`? I was trying to figure out if I can get rid of it but I have not gotten a response on [slack](https://dimagi.slack.com/archives/C0B1Z40DA/p1717765781969359) yet

MartinRiese commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -377,6 +378,8 @@ def extra_filters(self, for_count=False):
         if self.domain:
             filters.append(Q(domain=self.domain))
+        if self.not_domains:
+            filters.append(Q(domain__in=self.not_domains, _negated=True))
```
I just looked at the constructor for Q. `~` seems cleaner

MartinRiese commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -129,7 +137,8 @@ def print_data_row(data_row):
             for data_model in data_models:
                 try:
-                    process_data_model_fn = DATA_MODEL_HELPERS[data_model.lower()]()
+                    process_data_model_fn = DATA_MODEL_HELPERS[data_model.lower()](
+                    )
```
yes. reverted

MartinRiese committed:
Check one index at a time in stale_data_in_es

Instead of checking all ES indices on the same chuck, do one at a time.
This is to simplify the code since we only need to make a domain level
decision for some of the indices.

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -154,19 +164,47 @@ class CaseHelper:
+                yield from cls._yield_missing_in_es(chunk, case_search_query)
 
+    @staticmethod
+    def get_sql_chunks(domain, run_config, domains_with_dedicated_index=None):
```
`domains_with_dedicated_index` is unused now, right?

MartinRiese replied:
correct. Removed.

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -154,19 +164,47 @@ class CaseHelper:
+            case_search_index = settings.CASE_SEARCH_SUB_INDICES[domain]['index_cname']
+            case_search_query = CaseSearchES(for_export=True, index=case_search_index)
+            for chunk in cls.get_sql_chunks(domain, run_config):
+                yield from cls._yield_missing_in_es(chunk, case_search_query)
```
I like the clearer parallel between these three blocks of code, but unfortunately it means lookup up every document at least twice.  What if you inverted the for loops so you only need to fetch each chunk once from SQL, like this:

```python
domains = [run_config.domain] if run_config.domain is not ALL_DOMAINS else Domain.get_all_names()
domains_with_search_index = ...
domains_with_dedicated_index = ...
for domain in domains:
    for chunk in cls.get_sql_chunks(domain, run_config):
        case_query = CaseES(for_export=True)
        yield from cls._yield_missing_in_es(chunk, case_query)

        if domain in domains_with_search_index:
            case_search_query = CaseSearchES(for_export=True)
            yield from cls._yield_missing_in_es(chunk, case_search_query)

        if domain in domains_with_dedicated_index:
            case_search_index = settings.CASE_SEARCH_SUB_INDICES[domain]['index_cname']
            case_search_query = CaseSearchES(for_export=True, index=case_search_index)
            yield from cls._yield_missing_in_es(chunk, case_search_query)
```

MartinRiese replied:
Sounds good. I also added the do main to the iteration key so it is unique to the query.

MartinRiese committed:
Actually set the index

* Looks like you cannot set the index in the CaseSearchES constructor
  Call `query.index =` instead
* Use different iteration key for each query
* Revert changes to CaseReindexAccessor since they are not needed
  anymore

MartinRiese committed:
Use keys in setting directly

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

MartinRiese committed:
Inverse Loops

MartinRiese committed:
Fix tests

stale_data_in_es now returns two rows one for cases and one for
case_search

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -181,30 +181,30 @@ def run(cls, run_config):
 
         for domain in all_domains_with_dedicated_search_index:
             case_search_index = settings.CASE_SEARCH_SUB_INDICES[domain]['index_cname']
-            case_search_query = CaseSearchES(for_export=True, index=case_search_index)
```
This should work - it's what we do for case search: https://github.com/dimagi/commcare-hq/blob/29b3ff8a9890b4c4b832cb7974e1281c00202e5d/corehq/apps/case_search/utils.py#L181-L185
what was the error you were getting?

MartinRiese replied:
It was just not updating the index. Further down I was using it in `es_base_query.index` and if I set the index in the constructor I get "case_search" back instead of "case_search_bha"

esoergel replied:
Ah, I see - it doesn't override the `index` property, but it does set the adapter appropriately:
```python
>>> query = CaseSearchES(index='case_search_bha')
>>> query.index
'case_search'
>>> query.adapter.canonical_name
'case_search_bha'
```
So queries should go to the new index correctly, even though `query.index` isn't updated.  I'm not sure why we don't also set `index` there, that seems confusing/wrong.

esoergel commented on corehq/apps/hqadmin/management/commands/stale_data_in_es.py:
```diff
@@ -181,30 +181,30 @@ def run(cls, run_config):
         case_search_query = CaseSearchES(for_export=True)
         for domain in all_domains_with_search_index:
-            for chunk in cls.get_sql_chunks(domain, run_config):
+            for chunk in cls.get_sql_chunks(domain, run_config, f'case-search-{domain}'):
```
looks like `domain` is already in the key, for what it's worth.

esoergel replied:
(looks like this was removed anyways)
